zarr/_storage/store.py

```diff
@@ -303,6 +373,150 @@ def _ensure_store(store):
         )
 
 
+class StorageTransformer(MutableMapping, abc.ABC):
+    """Base class for storage transformers. The methods simply pass on the data as-is
+    and should be overwritten by sub-classes."""
+
+    _store_version = 3
+    _metadata_class = Metadata3
+
+    def __init__(self, _type) -> None:
+        if _type not in self.valid_types:  # pragma: no cover
+            raise ValueError(
+                f"Storage transformer cannot be initialized with type {_type}, "
+                + f"must be one of {list(self.valid_types)}."
+            )
+        self.type = _type
+        self._inner_store = None
+
+    def _copy_for_array(self, array, inner_store):
```

Comment:
what is the purpose of the unused `array` argument here? Is it a intended that subclasses would use this?

Reply:
Exactly, I've seen a need for this when implementing the Sharding storage transformer in #1111:
https://github.com/zarr-developers/zarr-python/blob/6e2790c6b5596471bbb47ccdcad8d9f035942f58/zarr/_storage/v3_storage_transformers.py#L99-L102

Related to this: I'm not super happy about the double-initialization using both `__init__` and `_copy_for_array`, but it seems necessary to supply the information in two steps, first from the config, then from the array. Please let me know if you have an idea how to solve this more elegantly.